File: afterfall-python/afterfall/render/world_view.py
"""Turns the simulation's world into Panda3D geometry and keeps the dynamic bits (homestead, felled trees,
fire, blood) in sync."""
import math

import numpy as np
from panda3d.core import LineSegs, NodePath, Quat, Vec3

from ..config import CHUNK, HALF, HN
from ..world import Pal, hexcol
from .geometry import make_geom_node, node_from_parts, to_panda, unit_node


def P(x, y, z):
    """Simulation point -> Panda3D point."""
    return Vec3(x, z, y)


class WorldView:
    def __init__(self, sim, root: NodePath):
        self.sim, self.world, self.home = sim, sim.world, sim.home
        self.root = root.attach_new_node("world")
        self._terrain()
        self._roads()
        self._water()
        self.static_root = self.root.attach_new_node("static")
        self._chunked(self.world.static_parts, self.static_root, low=False)
        self.veg_root = self.root.attach_new_node("vegetation")
        self.veg_count = 0
        self._rebuild_veg()
        self.tree_root = self.root.attach_new_node("trees")
        self.tree_nodes = {}
        by_chunk = {}
        for t in self.world.trees:
            by_chunk.setdefault(t.chunk, []).append(t)
        for key in by_chunk:
            self._build_tree_chunk(key)
        self.home_root = self.root.attach_new_node("homestead")
        self.project_nodes = {}
        self.unit = {s: NodePath(unit_node(s, low=True)) for s in ("cube", "sphere", "cylinder", "cone")}
        self._fire_props()
        self.falling, self.decals, self.tracers = [], [], []
        self.fx_seen = 0

    # --- static geometry -------------------------------------------------------------------------

    def _terrain(self):
        w = self.world
        s = HN + 1
        pos = np.stack([w.gx, w.gy, w.h], axis=-1).reshape(-1, 3)
        nrm = w.normals.reshape(-1, 3)
        col = np.concatenate([w.colors.reshape(-1, 3), np.zeros((s * s, 1))], axis=1)
        i, j = np.meshgrid(np.arange(HN), np.arange(HN))
        a = (j * s + i).ravel()
        tris = np.concatenate([np.stack([a, a + 1, a + s + 1], 1), np.stack([a, a + s + 1, a + s], 1)])
        self.root.attach_new_node(make_geom_node("terrain", to_panda(pos), to_panda(nrm), col, tris))

    def _roads(self):
        w = self.world
        self.road_nodes = []
        for road in w.roads:
            sm = road["samples"]
            width = road["w"]
            base = np.array(hexcol(Pal.ASPHALT if road["asphalt"] else Pal.DIRT_ROAD))
            pos, cols = [], []
            for i in range(len(sm)):
                t = sm[min(len(sm) - 1, i + 1)] - sm[max(0, i - 1)]
                t = t / (np.linalg.norm(t) + 1e-9)
                n = np.array([-t[1], t[0]])
                for k in range(5):
                    off = (k / 4 - .5) * width
                    x, y = sm[i] + n * off
                    edge = abs(k - 2) / 2
                    pos.append((x, y, w.ground_z(x, y) + .07 + (.02 if k == 2 else 0)))
                    shade = 1 - .12 * edge + ((i * 7 + k * 3) % 5) * .012
                    cols.append((*(base * shade), 0))
            pos = np.array(pos)
            nrm = np.tile([0, 0, 1.0], (len(pos), 1))
            tris = []
            for i in range(len(sm) - 1):
                for k in range(4):
                    a = i * 5 + k
                    tris += [(a, a + 5, a + 6), (a, a + 6, a + 1)]
            node = self.root.attach_new_node(make_geom_node("road", to_panda(pos), to_panda(nrm), np.array(cols), np.array(tris)))
            self.road_nodes.append(node)
            node.set_shader_input("u_gloss", 0.0)

    def _water(self):
        w = self.world
        px, py, pr = w.pond
        seg, r = 64, pr + 16
        pos = [(px, py, w.water_z)] + [(px + math.cos(a) * r, py + math.sin(a) * r, w.water_z) for a in np.linspace(0, 2 * math.pi, seg, endpoint=False)]
        pos = np.array(pos)
        tris = np.array([(0, 1 + k, 1 + (k + 1) % seg) for k in range(seg)])
        nrm = np.tile([0, 0, 1.0], (len(pos), 1))
        col = np.tile([*hexcol(Pal.WATER), 0], (len(pos), 1))
        self.water = self.root.attach_new_node(make_geom_node("water", to_panda(pos), to_panda(nrm), col, tris))
        self.water.set_shader_input("u_gloss", .9)

    def _chunked(self, parts, parent, low):
        groups = {}
        for p in parts:
            groups.setdefault((int((p.pos[0] + HALF) // CHUNK), int((p.pos[1] + HALF) // CHUNK)), []).append(p)
        for key, group in groups.items():
            node = node_from_parts(f"chunk{key}", group, low)
            if node:
                parent.attach_new_node(node)

    def _rebuild_veg(self):
        self.veg_root.node().remove_all_children()
        self._chunked(self.world.veg_parts, self.veg_root, low=True)
        self.veg_count = len(self.world.veg_parts)

    def _build_tree_chunk(self, key):
        old = self.tree_nodes.pop(key, None)
        if old is not None:
            old.remove_node()
        parts = [p for t in self.world.trees if t.chunk == key and t.alive for p in t.parts]
        node = node_from_parts(f"trees{key}", parts, low=True)
        if node:
            self.tree_nodes[key] = self.tree_root.attach_new_node(node)

    # --- homestead ---------------------------------------------------------------------------------

    def _project_node(self, key):
        p = self.home.by_key[key]
        parts = self.home.build_parts(key)
        if key.startswith("pal"):
            ax, ay = self.home.base.x, self.home.base.y
        else:
            ax, ay = self.home.pos(p.at)
        az = self.world.ground_z(ax, ay)
        pivot = self.home_root.attach_new_node(f"proj_{key}")
        pivot.set_pos(P(ax, ay, az))
        geom = node_from_parts(key, parts, low=False)
        if geom:
            child = pivot.attach_new_node(geom)
            child.set_pos(-P(ax, ay, az))
        return pivot

    def _fire_props(self):
        fx, fy = self.home.fire_pos
        fz = self.world.ground_z(fx, fy)
        self.fire_pos = P(fx, fy, fz)
        self.flames = self.home_root.attach_new_node("flames")
        self.flames.set_pos(self.fire_pos)
        self.flame_parts = []
        for i, (c, s) in enumerate(((0xffb347, 1.0), (0xff7a1f, .8), (0xffe08a, .55))):
            n = self.flames.attach_new_node("flame")
            self.unit["cone"].instance_to(n)
            n.set_color_scale(*hexcol(c), 1)
            n.set_shader_input("u_emissive", 1.6)
            n.set_pos(math.cos(i * 2.1) * .1, .45, math.sin(i * 2.1) * .1)
            n.set_scale(.5 * s, .9 * s, .5 * s)
            self.flame_parts.append((n, s))
        self.plants = self.home_root.attach_new_node("plants")
        for g in ("garden1", "garden2"):
            p = self.home.by_key[g]
            gx, gy = self.home.pos(p.at)
            gz = self.world.ground_z(gx, gy)
            for r in range(4):
                for c in range(7):
                    n = self.plants.attach_new_node(f"plant_{g}")
                    self.unit["sphere"].instance_to(n)
                    n.set_pos(P(gx - 1.8 + c * .6, gy - 1.1 + r * .73, gz + .3))
                    n.set_scale(.36, .26, .36)
                    n.set_color_scale(*(hexcol(0x86a55a) if r % 2 else hexcol(0x4f7a2e)), 1)
                    n.set_tag("garden", g)
        self.flag = None
        self.lamp = None
        self.radio_lamp = None

    def update_home(self):
        home = self.home
        wanted = set(home.built) | set(home.progress)
        for key in wanted:
            if key not in self.project_nodes:
                self.project_nodes[key] = self._project_node(key)
        for key, node in list(self.project_nodes.items()):
            if key not in wanted:
                node.remove_node()
                del self.project_nodes[key]
                continue
            prog = 1.0 if key in home.built else home.progress.get(key, 0.0)
            node.set_scale(1, max(.02, prog * prog * (3 - 2 * prog)), 1)
        if home.breached is not None and f"pal{home.breached}" in self.project_nodes:
            self.project_nodes[f"pal{home.breached}"].set_scale(1, .25, 1)
        grow = .35 + .65 * min(1.0, home.garden_food / 4)
        for n in self.plants.get_children():
            if n.get_tag("garden") in home.built:
                n.set_scale(.36 * grow, .26 * grow, .36 * grow)
            else:
                n.set_scale(.001)
        if "flag" in home.built and self.flag is None:
            fx, fy = home.pos(home.by_key["flag"].at)
            fz = self.world.ground_z(fx, fy)
            self.flag = self.home_root.attach_new_node("flag")
            self.flag.set_pos(P(fx, fy, fz + 6.3))
            cloth = self.flag.attach_new_node("cloth")
            self.unit["cube"].instance_to(cloth)
            cloth.set_pos(.9, 0, 0)
            cloth.set_scale(1.8, 1.1, .03)
            cloth.set_color_scale(*hexcol(0xd8d1bd), 1)
            emblem = self.flag.attach_new_node("emblem")
            self.unit["cylinder"].instance_to(emblem)
            emblem.set_pos(.9, 0, 0)
            emblem.set_hpr(0, 90, 0)
            emblem.set_scale(.7, .05, .7)
            emblem.set_color_scale(*hexcol(0x2f4a2c), 1)
        if "gen" in home.built and self.lamp is None:
            gx, gy = home.pos(home.by_key["gen"].at)
            self.lamp_pos = P(gx + 1.6, gy, self.world.ground_z(gx, gy) + 5.9)
            self.lamp = True
        if "radio" in home.built and self.radio_lamp is None:
            rx, ry = home.pos(home.by_key["radio"].at)
            self.radio_lamp = self.home_root.attach_new_node("radio_lamp")
            self.unit["sphere"].instance_to(self.radio_lamp)
            self.radio_lamp.set_pos(P(rx, ry, self.world.ground_z(rx, ry) + 15.2))
            self.radio_lamp.set_scale(.25)
            self.radio_lamp.set_color_scale(1, .15, .1, 1)
        home.changed = False

    # --- per-frame ---------------------------------------------------------------------------------

    def update(self, t, dt, night):
        home = self.home
        if home.changed or home.progress:
            self.update_home()
        fire_on = "fire" in home.built and home.fire_fuel > 0
        self.flames.set_scale(1 if fire_on else .001)
        for i, (n, s) in enumerate(self.flame_parts):
            f = .8 + .25 * math.sin(t * 11 + i * 2) * math.sin(t * 7.3 + i)
            n.set_scale(.5 * s * (1 + .1 * math.sin(t * 9 + i)), .9 * s * f, .5 * s)
        if self.flag is not None:
            self.flag.set_hpr(math.sin(t * 1.3) * 25 + 40, 0, math.sin(t * 5) * 4)
        if self.radio_lamp is not None:
            self.radio_lamp.set_shader_input("u_emissive", 4.0 if math.sin(t * 3) > .6 else 0.0)
        # felled trees and new stumps
        for key in list(self.world.dirty_chunks):
            self._build_tree_chunk(key)
        self.world.dirty_chunks.clear()
        if len(self.world.veg_parts) != self.veg_count:
            self._rebuild_veg()
        self._effects(dt)
        return fire_on

    def fire_light(self, t, night):
        home = self.home
        if "fire" not in home.built or home.fire_fuel <= 0:
            return (0, 0, 0, 0), (0, 0, 0)
        f = .8 + .2 * math.sin(t * 23) * math.sin(t * 7.3)
        p = self.fire_pos + Vec3(0, 1.1, 0)
        k = (.35 + night * 1.4) * f
        return (p.x, p.y, p.z, 20.0), (1.0 * k, .5 * k, .18 * k)

    def flood_light(self, night):
        if self.lamp is None or night < .4:
            return (0, 0, 0, 0), (0, 0, 0)
        p = self.lamp_pos
        return (p.x, p.y, p.z, 32.0), (1.1, 1.0, .85)

    def _effects(self, dt):
        sim = self.sim
        new = sim.fx[self.fx_seen:]
        self.fx_seen = len(sim.fx)
        for fx in new:
            if fx[0] == "fell":
                self._start_fall(self.world.trees[fx[1]], fx[2])
            elif fx[0] == "blood":
                x, y = fx[1], fx[2]
                n = self.root.attach_new_node("blood")
                self.unit["cylinder"].instance_to(n)
                n.set_pos(P(x, y, self.world.ground_z(x, y) + .03))
                n.set_scale(1.3, .02, 1.3)
                n.set_color_scale(.22, .02, .01, 1)
                self.decals.append([n, 0.0])
            elif fx[0] == "shot":
                _, sx, sy, zx, zy, hit = fx
                ls = LineSegs()
                ls.set_color(1, .88, .6, 1)
                ls.set_thickness(2)
                ls.move_to(P(sx, sy, self.world.ground_z(sx, sy) + 1.45))
                ls.draw_to(P(zx, zy, self.world.ground_z(zx, zy) + (1.5 if hit else 2.5)))
                n = self.root.attach_new_node(ls.create())
                n.set_shader_off()
                self.tracers.append([n, 0.0])
        if len(sim.fx) > 400:
            sim.fx = sim.fx[-100:]
            self.fx_seen = len(sim.fx)
        for item in self.falling[:]:
            node, axis, t = item
            item[2] += dt
            k = min(1.0, (item[2] / 2.2) ** 2)
            q = Quat()
            q.set_from_axis_angle(k * 88, axis)
            node.set_quat(q)
            if item[2] > 14:
                node.set_z(node.get_z() - dt * .3)
            if item[2] > 22:
                node.remove_node()
                self.falling.remove(item)
        for item in self.decals[:]:
            item[1] += dt
            if item[1] > 90:
                item[0].remove_node()
                self.decals.remove(item)
        for item in self.tracers[:]:
            item[1] += dt
            if item[1] > .15:
                item[0].remove_node()
                self.tracers.remove(item)

    def _start_fall(self, tree, direction):
        geom = node_from_parts("falling", tree.parts, low=True)
        if not geom:
            return
        base = P(tree.x, tree.y, self.world.ground_z(tree.x, tree.y))
        pivot = self.root.attach_new_node("fall")
        pivot.set_pos(base)
        child = pivot.attach_new_node(geom)
        child.set_pos(-base)
        d = Vec3(math.cos(direction), 0, math.sin(direction))
        axis = Vec3(0, 1, 0).cross(d)
        axis.normalize()
        self.falling.append([pivot, -axis, 0.0])

File: afterfall-unity/Assets/Afterfall/Runtime/WorldView.cs
// Builds the Unity scene for the simulated world and keeps its dynamic parts in sync: felled trees,
// homestead projects rising as they're built, the campfire and floodlight, blood, tracers.
using System.Collections.Generic;
using System.Linq;
using Afterfall.Sim;
using UnityEngine;
using UnityEngine.Rendering;
using UnityEngine.Rendering.HighDefinition;
using static Afterfall.Game.MeshBuilder;

namespace Afterfall.Game
{
    public sealed class WorldView
    {
        readonly Simulation sim;
        readonly World world;
        readonly Homestead home;
        readonly MaterialLibrary lib;
        readonly AfterfallAssets assets;
        public readonly Transform root;
        public Terrain terrain;
        Transform vegRoot, treeRoot, homeRoot;
        readonly Dictionary<(int, int), GameObject> treeChunks = new Dictionary<(int, int), GameObject>();
        readonly Dictionary<string, Transform> projects = new Dictionary<string, Transform>();
        int vegCount;
        Transform flames;
        readonly List<(Transform t, float s)> flameParts = new List<(Transform, float)>();
        readonly List<(Transform t, string garden)> plants = new List<(Transform, string)>();
        Light fireLight, floodLight;
        HDAdditionalLightData fireHd;
        Transform flag, radioLamp;
        Material radioLampMat;
        readonly List<(Transform t, float age)> decals = new List<(Transform, float)>();
        readonly List<(GameObject g, float age)> tracers = new List<(GameObject, float)>();
        readonly List<(Transform pivot, Vector3 dir, float t)> falling = new List<(Transform, Vector3, float)>();
        int fxSeen;
        Material bloodMat, tracerMat;

        public Vector3 FirePos { get; private set; }

        public WorldView(Simulation sim, MaterialLibrary lib, AfterfallAssets assets, bool grass)
        {
            this.sim = sim;
            world = sim.world;
            home = sim.home;
            this.lib = lib;
            this.assets = assets;
            root = new GameObject("World").transform;
            terrain = TerrainBuilder.Build(world, assets, lib, root, grass);
            Water();
            var stat = new GameObject("Static").transform;
            stat.SetParent(root, false);
            Chunked(world.staticParts, stat, false);
            vegRoot = new GameObject("Vegetation").transform;
            vegRoot.SetParent(root, false);
            RebuildVeg();
            treeRoot = new GameObject("Trees").transform;
            treeRoot.SetParent(root, false);
            foreach (var key in world.trees.Select(t => t.chunk).Distinct()) BuildTreeChunk(key);
            homeRoot = new GameObject("Homestead").transform;
            homeRoot.SetParent(root, false);
            FireProps();
            bloodMat = lib.NewLit();
            bloodMat.SetColor("_BaseColor", new Color(.2f, .015f, .01f));
            bloodMat.SetFloat("_Smoothness", .7f);
            MaterialLibrary.Validate(bloodMat);
            tracerMat = lib.Emissive(new Color(1, .85f, .55f), 4000);
        }

        public float Ground(double x, double y) => terrain != null ? terrain.SampleHeight(U(x, y, 0)) + terrain.transform.position.y : (float)world.GroundZ(x, y);

        void Water()
        {
            var (px, py, pr) = world.pond;
            var go = GameObject.CreatePrimitive(PrimitiveType.Cylinder);
            Object.Destroy(go.GetComponent<Collider>());
            go.name = "Pond";
            go.transform.SetParent(root, false);
            go.transform.position = U(px, py, world.waterZ);
            go.transform.localScale = new Vector3((float)(pr + 16) * 2, .01f, (float)(pr + 16) * 2);
            var m = lib.Get(null, Pal.WATER);
            go.GetComponent<MeshRenderer>().sharedMaterial = m;
        }

        void Chunked(List<Part> parts, Transform parent, bool low)
        {
            foreach (var g in parts.GroupBy(p => ((int)((p.pos.x + Config.HALF) / Config.CHUNK), (int)((p.pos.y + Config.HALF) / Config.CHUNK))))
                MeshBuilder.Create($"chunk{g.Key}", g.ToList(), lib, parent, low);
        }

        void RebuildVeg()
        {
            foreach (Transform c in vegRoot) Object.Destroy(c.gameObject);
            Chunked(world.vegParts, vegRoot, true);
            vegCount = world.vegParts.Count;
        }

        void BuildTreeChunk((int, int) key)
        {
            if (treeChunks.TryGetValue(key, out var old)) Object.Destroy(old);
            var parts = world.trees.Where(t => t.chunk == key && t.alive).SelectMany(t => t.parts).ToList();
            var go = MeshBuilder.Create($"trees{key}", parts, lib, treeRoot, true);
            if (go != null) treeChunks[key] = go;
            else treeChunks.Remove(key);
        }

        // --- homestead -------------------------------------------------------------------------------

        Transform ProjectNode(string key)
        {
            var p = home.byKey[key];
            var at = key.StartsWith("pal") ? new V2(home.baseSite.x, home.baseSite.y) : home.Pos(p.at);
            var pivotPos = U(at.x, at.y, world.GroundZ(at.x, at.y));
            var pivot = new GameObject("proj_" + key).transform;
            pivot.SetParent(homeRoot, false);
            pivot.position = pivotPos;
            // parts are in world space: build them relative to the pivot, which sits at the project's foot
            var go = MeshBuilder.Create(key, home.BuildParts(key), lib, pivot, false, true, pivotPos);
            if (go != null) go.transform.localPosition = Vector3.zero;
            return pivot;
        }

        Light AddLight(string name, LightType type, float range, Color color, bool shadows)
        {
            var go = new GameObject(name);
            go.transform.SetParent(homeRoot, false);
            var l = go.AddComponent<Light>();
            l.type = type;
            var hd = go.AddComponent<HDAdditionalLightData>();
            l.range = range;
            l.color = color;
            l.lightUnit = LightUnit.Lumen;
            l.shadows = shadows ? LightShadows.Soft : LightShadows.None;
            if (shadows) hd.SetShadowResolution(512);
            return l;
        }

        void FireProps()
        {
            var fp = home.FirePos;
            FirePos = U(fp.x, fp.y, world.GroundZ(fp.x, fp.y));
            flames = new GameObject("Flames").transform;
            flames.SetParent(homeRoot, false);
            flames.position = FirePos;
            var cols = new[] { (new Color(1, .7f, .28f), 1f), (new Color(1, .48f, .12f), .8f), (new Color(1, .88f, .54f), .55f) };
            for (int i = 0; i < cols.Length; i++)
            {
                var (c, s) = cols[i];
                var part = new Part { shape = Shape.Cone, m = Mat3.Identity, pos = new V3(0, 0, 0), color = 0 };
                var (mesh, _) = MeshBuilder.Build(new[] { part }, lib);
                var go = new GameObject("flame");
                go.transform.SetParent(flames, false);
                go.AddComponent<MeshFilter>().sharedMesh = mesh;
                var r = go.AddComponent<MeshRenderer>();
                r.sharedMaterial = lib.Emissive(c, 9000 * s);
                r.shadowCastingMode = ShadowCastingMode.Off;
                go.transform.localPosition = new Vector3(Mathf.Cos(i * 2.1f) * .1f, .45f, Mathf.Sin(i * 2.1f) * .1f);
                flameParts.Add((go.transform, s));
            }
            fireLight = AddLight("FireLight", LightType.Point, 22, new Color(1, .55f, .25f), true);
            fireLight.transform.position = FirePos + Vector3.up * 1.1f;
            fireHd = fireLight.GetComponent<HDAdditionalLightData>();
            var leaf = MaterialLibrary.Hex(0x4f7a2e);
            foreach (var g in new[] { "garden1", "garden2" })
            {
                var at = home.Pos(home.byKey[g].at);
                float gz = (float)world.GroundZ(at.x, at.y);
                for (int r = 0; r < 4; r++)
                    for (int c = 0; c < 7; c++)
                    {
                        var s = GameObject.CreatePrimitive(PrimitiveType.Sphere);
                        Object.Destroy(s.GetComponent<Collider>());
                        s.transform.SetParent(homeRoot, false);
                        s.transform.position = U(at.x - 1.8 + c * .6, at.y - 1.1 + r * .73, gz + .3);
                        s.GetComponent<MeshRenderer>().sharedMaterial = lib.Get("leaves", r % 2 == 1 ? 0x6b8a34 : 0x44602c);
                        s.transform.localScale = Vector3.one * .001f;
                        plants.Add((s.transform, g));
                    }
            }
            _ = leaf;
        }

        void UpdateHome()
        {
            var wanted = new HashSet<string>(home.built.Concat(home.progress.Keys));
            foreach (var key in wanted)
                if (!projects.ContainsKey(key)) projects[key] = ProjectNode(key);
            foreach (var key in projects.Keys.ToList())
            {
                if (!wanted.Contains(key))
                {
                    Object.Destroy(projects[key].gameObject);
                    projects.Remove(key);
                    continue;
                }
                float prog = home.built.Contains(key) ? 1 : (float)(home.progress.TryGetValue(key, out double v) ? v : 0);
                projects[key].localScale = new Vector3(1, Mathf.Max(.02f, prog * prog * (3 - 2 * prog)), 1);
            }
            if (home.breached.HasValue && projects.TryGetValue("pal" + home.breached.Value, out var broken)) broken.localScale = new Vector3(1, .25f, 1);
            float grow = .35f + .65f * Mathf.Min(1, home.gardenFood / 4f);
            foreach (var (t, g) in plants)
                t.localScale = home.built.Contains(g) ? new Vector3(.36f, .26f, .36f) * grow : Vector3.one * .001f;
            if (home.built.Contains("flag") && flag == null)
            {
                var at = home.Pos(home.byKey["flag"].at);
                flag = new GameObject("Flag").transform;
                flag.SetParent(homeRoot, false);
                flag.position = U(at.x, at.y, world.GroundZ(at.x, at.y) + 6.3);
                var cloth = GameObject.CreatePrimitive(PrimitiveType.Cube);
                Object.Destroy(cloth.GetComponent<Collider>());
                cloth.transform.SetParent(flag, false);
                cloth.transform.localPosition = new Vector3(.9f, 0, 0);
                cloth.transform.localScale = new Vector3(1.8f, 1.1f, .03f);
                cloth.GetComponent<MeshRenderer>().sharedMaterial = lib.Get("fabric", 0xd8d1bd);
            }
            if (home.built.Contains("gen") && floodLight == null)
            {
                var at = home.Pos(home.byKey["gen"].at);
                floodLight = AddLight("Floodlight", LightType.Spot, 45, new Color(1, .95f, .85f), true);
                floodLight.spotAngle = 95;
                floodLight.transform.position = U(at.x + 1.6, at.y, world.GroundZ(at.x, at.y) + 5.9);
                floodLight.transform.rotation = Quaternion.LookRotation(U(home.baseSite.x, home.baseSite.y, 0) - U(at.x + 1.6, at.y, 0) + Vector3.down * 5);
            }
            if (home.built.Contains("radio") && radioLamp == null)
            {
                var at = home.Pos(home.byKey["radio"].at);
                var s = GameObject.CreatePrimitive(PrimitiveType.Sphere);
                Object.Destroy(s.GetComponent<Collider>());
                radioLamp = s.transform;
                radioLamp.SetParent(homeRoot, false);
                radioLamp.position = U(at.x, at.y, world.GroundZ(at.x, at.y) + 15.2);
                radioLamp.localScale = Vector3.one * .25f;
                radioLampMat = lib.Emissive(new Color(1, .15f, .1f), 20000);
                s.GetComponent<MeshRenderer>().sharedMaterial = radioLampMat;
            }
            home.changed = false;
        }

        // --- per frame -----------------------------------------------------------------------------------

        public void Update(float t, float dt, float night)
        {
            if (home.changed || home.progress.Count > 0) UpdateHome();
            bool fireOn = home.built.Contains("fire") && home.fireFuel > 0;
            flames.gameObject.SetActive(fireOn);
            for (int i = 0; i < flameParts.Count; i++)
            {
                var (n, s) = flameParts[i];
                float f = .8f + .25f * Mathf.Sin(t * 11 + i * 2) * Mathf.Sin(t * 7.3f + i);
                n.localScale = new Vector3(.5f * s * (1 + .1f * Mathf.Sin(t * 9 + i)), .9f * s * f, .5f * s);
            }
            fireLight.enabled = fireOn;
            if (fireOn) fireLight.intensity = 5200 * (.8f + .2f * Mathf.Sin(t * 23) * Mathf.Sin(t * 7.3f));
            if (floodLight != null)
            {
                floodLight.enabled = night > .4f;
                floodLight.intensity = 9000;
            }
            if (flag != null) flag.localRotation = Quaternion.Euler(Mathf.Sin(t * 5) * 4, Mathf.Sin(t * 1.3f) * 25 + 40, 0);
            if (radioLamp != null) radioLamp.gameObject.SetActive(Mathf.Sin(t * 3) > .6f);
            foreach (var key in world.dirtyChunks) BuildTreeChunk(key);
            world.dirtyChunks.Clear();
            if (world.vegParts.Count != vegCount) RebuildVeg();
            Effects(dt);
        }

        void Effects(float dt)
        {
            for (int i = fxSeen; i < sim.fx.Count; i++)
            {
                var f = sim.fx[i];
                if (f.kind == "fell") StartFall(world.trees[f.tree], f.angle);
                else if (f.kind == "blood")
                {
                    var d = GameObject.CreatePrimitive(PrimitiveType.Cylinder);
                    Object.Destroy(d.GetComponent<Collider>());
                    d.transform.SetParent(root, false);
                    d.transform.position = new Vector3((float)f.x, Ground(f.x, f.y) + .02f, (float)f.y);
                    d.transform.localScale = new Vector3(1.3f, .01f, 1.3f) * Random.Range(.7f, 1.2f);
                    d.GetComponent<MeshRenderer>().sharedMaterial = bloodMat;
                    d.GetComponent<MeshRenderer>().shadowCastingMode = ShadowCastingMode.Off;
                    decals.Add((d.transform, 0));
                }
                else if (f.kind == "shot")
                {
                    var g = new GameObject("tracer");
                    g.transform.SetParent(root, false);
                    var lr = g.AddComponent<LineRenderer>();
                    lr.sharedMaterial = tracerMat;
                    lr.widthMultiplier = .03f;
                    lr.positionCount = 2;
                    lr.SetPosition(0, new Vector3((float)f.x, Ground(f.x, f.y) + 1.45f, (float)f.y));
                    lr.SetPosition(1, new Vector3((float)f.x2, Ground(f.x2, f.y2) + (f.hit ? 1.5f : 2.5f), (float)f.y2));
                    tracers.Add((g, 0));
                }
            }
            fxSeen = sim.fx.Count;
            if (sim.fx.Count > 400)
            {
                sim.fx.RemoveRange(0, sim.fx.Count - 100);
                fxSeen = sim.fx.Count;
            }
            for (int i = falling.Count - 1; i >= 0; i--)
            {
                var (pivot, axis, t0) = falling[i];
                float t = t0 + dt;
                float k = Mathf.Min(1, (t / 2.2f) * (t / 2.2f));
                pivot.localRotation = Quaternion.Slerp(Quaternion.identity, Quaternion.FromToRotation(Vector3.up, axis), k);
                if (t > 14) pivot.position += Vector3.down * dt * .3f;
                if (t > 22) { Object.Destroy(pivot.gameObject); falling.RemoveAt(i); }
                else falling[i] = (pivot, axis, t);
            }
            for (int i = decals.Count - 1; i >= 0; i--)
            {
                var (t, age) = decals[i];
                if (age + dt > 90) { Object.Destroy(t.gameObject); decals.RemoveAt(i); }
                else decals[i] = (t, age + dt);
            }
            for (int i = tracers.Count - 1; i >= 0; i--)
            {
                var (g, age) = tracers[i];
                if (age + dt > .15f) { Object.Destroy(g); tracers.RemoveAt(i); }
                else tracers[i] = (g, age + dt);
            }
        }

        void StartFall(Afterfall.Sim.Tree tree, double direction)
        {
            var basePos = U(tree.x, tree.y, world.GroundZ(tree.x, tree.y));
            var pivot = new GameObject("falling").transform;
            pivot.SetParent(root, false);
            pivot.position = basePos;
            var go = MeshBuilder.Create("tree", tree.parts, lib, pivot, true, true, basePos);
            if (go == null) { Object.Destroy(pivot.gameObject); return; }
            go.transform.localPosition = Vector3.zero;
            // falls away from the survivor (simulation angle -> Unity x/z)
            var d = new Vector3(Mathf.Cos((float)direction), .035f, Mathf.Sin((float)direction)).normalized;
            falling.Add((pivot, d, 0));
        }
    }
}
